### PR TITLE
[Core] Improve table performance

### PR DESCRIPTION
**Pull request by jrubiogonzalez**
We realized that on some of our calculations accessing the table values was one of the main bottlenecks. This PR tries to smoothen them by:

1. Firs checks if the value is out of bounds, so for the search it is inside the table range.
2.  Improving the lower and upper values search for the interpolation adding a binary search. Includes some tricks for trying to improve performance.
3. Improve the Interpolation calculation to improve performance, now it is formulated without if's.
4. Fix a bug I noticed when computing derivatives.

### CODE REVIEW

File: kratos/includes/table.h [hunk 1]
```diff
@@ -521,23 +521,30 @@ class Table<double, double>
     // Get the value for the given argument using piecewise linear
     result_type GetValue(argument_type const& X) const
     {
-        std::size_t size = mData.size();
-
-        KRATOS_ERROR_IF(size == 0) << "Get value from empty table" << std::endl;
-
-        if(size==1) // constant table. Returning the only value we have.
+        int u = mData.size()-1; //using directly u for optimizing one allocation
```
Comment (loumalouomega) [inline kratos/includes/table.h:528]:
```suggestion
        const int u = mData.size()-1; //using directly u for optimizing one allocation
```

File: kratos/includes/table.h [hunk 2]
```diff
@@ -521,23 +521,30 @@ class Table<double, double>
...
-        if(X <= mData[0].first)
-            return Interpolate(X, mData[0].first, mData[0].second[0], mData[1].first, mData[1].second[0], result);
-
-        for(std::size_t i = 1 ; i < size ; i++)
-            if(X <= mData[i].first)
-                return Interpolate(X, mData[i-1].first, mData[i-1].second[0], mData[i].first, mData[i].second[0], result);
-
-        // now the x is outside the table and we hae to extrapolate it using last two records of table.
-        return Interpolate(X, mData[size-2].first, mData[size-2].second[0], mData[size-1].first, mData[size-1].second[0], result);
+        if(X <=  mData.begin()->first)
```
Comment (loumalouomega) [inline kratos/includes/table.h]: Add {} ¿?

File: kratos/includes/table.h [hunk 3]
```diff
@@ -660,30 +656,34 @@ class Table<double, double>
...
-
-        result_type result;
-        if(X <= mData[0].first)
-            //return Interpolate(X, mData[0].first, mData[0].second[0], mData[1].first, mData[1].second[0], result);
-            return 0.0;
-
-        for(std::size_t i = 1 ; i < size ; i++)
-            if(X <= mData[i].first)
-                return InterpolateDerivative( mData[i-1].first, mData[i-1].second[0], mData[i].first, mData[i].second[0], result);
+        int u = mData.size()-1; //using directly u for optimizing one allocation
```
Comment (loumalouomega) [inline kratos/includes/table.h:674]:
```suggestion
        const int u = mData.size()-1; //using directly u for optimizing one allocation
```

**Review by loumalouomega — COMMENTED** [review]
Minor comments

File: kratos/includes/table.h (hunk 1 above)
Comment (jrubiogonzalez) [reply kratos/includes/table.h:528]: no, because later on it might be changed in the binary search

**Comment by philbucher** [comment]
There seems to be only little testing for the table. If it makes sense, could you please add your testings also as unit tests?

**Comment by jrubiogonzalez** [comment]
I have added few more tests, now I think it covers all possible combinations. By the way @loumalouomega , ContactMechanicsApplicationTests was failing because some values added to a table were duplicated (two values for the same X) and trying to do extrapolation. Please check your tests.

**Comment by loumalouomega** [comment]
> I have added few more tests, now I think it covers all possible combinations. By the way @loumalouomega , ContactMechanicsApplicationTests was failing because some values added to a table were duplicated (two values for the same X) and trying to do extrapolation. Please check your tests.

Okidoki

**Comment by loumalouomega** [comment]
> I have added few more tests, now I think it covers all possible combinations. By the way @loumalouomega , ContactMechanicsApplicationTests was failing because some values added to a table were duplicated (two values for the same X) and trying to do extrapolation. Please check your tests.

imagen

Which tests?

**Comment by jrubiogonzalez** [comment]
I fixed the new implementation to keep the same old behavior even for strange input. But you can check them in
NightlyTests.ALMThreeDSimplestPatchMatchingAdaptativeTestContact and other (check https://github.com/KratosMultiphysics/Kratos/runs/4407342781?check_suite_focus=true )

The reason is that a table is used (inside results_database) that have the very last value duplicated. Now everything is working, but my understanding is that something has slipped in this test definition.

**Comment by loumalouomega** [comment]
> I fixed the new implementation to keep the same old behavior even for strange input. But you can check them in NightlyTests.ALMThreeDSimplestPatchMatchingAdaptativeTestContact and other (check [KratosMultiphysics/Kratos/runs/4407342781?check_suite_focus=true](https://github.com/KratosMultiphysics/Kratos/runs/4407342781?check_suite_focus=true) )
> 
> The reason is that a table is used (inside results_database) that have the very last value duplicated. Now everything is working, but my understanding is that something has slipped in this test definition.

That test is special, but I will check anyway

**Comment by loumalouomega** [comment]
> > I fixed the new implementation to keep the same old behavior even for strange input. But you can check them in NightlyTests.ALMThreeDSimplestPatchMatchingAdaptativeTestContact and other (check [KratosMultiphysics/Kratos/runs/4407342781?check_suite_focus=true](https://github.com/KratosMultiphysics/Kratos/runs/4407342781?check_suite_focus=true) )
> > The reason is that a table is used (inside results_database) that have the very last value duplicated. Now everything is working, but my understanding is that something has slipped in this test definition.
> 
> That test is special, but I will check anyway

(I an adaptive NR, so the times may be repeated)

**Comment by jrubiogonzalez** [comment]
@philbucher @loumalouomega somebody to approve?

**Review by loumalouomega — DISMISSED** [review]

File: kratos/includes/table.h [hunk 4]
```diff
@@ -603,24 +615,25 @@ class Table<double, double>
...
 
-        double scale = 0.00;
-
-        if (dx > epsilon)
-            scale = (X - X1) / dx;
+    }
+    result_type ExtrapolateBelowLowest(argument_type const& X, argument_type const& X1, result_type const& Y1, argument_type const& X2, result_type const& Y2) const
+    {
+        // const double epsilon = 1e-12;
+        return Y1 - (X2 - X1 > 1.0e-12) * (Y2 - Y1)*(X1 - X)/(X2 - X1 + std::numeric_limits<double>::min());
```
Comment (pooyan-dadvand) [inline kratos/includes/table.h]: some parenthesis would help the readability of the code

File: kratos/includes/table.h [hunk 5]
```diff
@@ -603,24 +615,25 @@ class Table<double, double>
...
 
-        Result = Y1 + dy * scale;
+    }
 
-        return Result;
+    result_type ExtrapolateAboveHighest(argument_type const& X, argument_type const& X1, result_type const& Y1, argument_type const& X2, result_type const& Y2) const
+    {
+        // const double epsilon = 1e-12;
+        return Y2 + (X2 - X1 > 1.0e-12) * (Y2 - Y1)*(X - X2)/(X2 - X1 + std::numeric_limits<double>::min());
```
Comment (pooyan-dadvand) [inline kratos/includes/table.h]: same here

File: kratos/includes/table.h [hunk 6]
```diff
@@ -660,30 +673,36 @@ class Table<double, double>
...
+            return InterpolateDerivative(mData.begin()->first, mData.begin()->second[0], mData[1].first, mData[1].second[0]);
+        }
 
-        result_type result;
-        if(X <= mData[0].first)
-            //return Interpolate(X, mData[0].first, mData[0].second[0], mData[1].first, mData[1].second[0], result);
-            return 0.0;
+        if(X >= mData[u].first){
+            // now the x is outside the table and we hae to extrapolate it using last two records of table.
+            return InterpolateDerivative(mData[u-1].first, mData[u-1].second[0], mData[u].first, mData[u].second[0]);
```
Comment (pooyan-dadvand) [inline kratos/includes/table.h:682]: I would change the comment saying that we use the last gradient for the extrapolation. It is not clear why you interpolate after saying you are extrapolating

File: kratos/includes/table.h [hunk 7]
```diff
@@ -603,24 +615,25 @@ class Table<double, double>
...
 
-    result_type& Interpolate(argument_type const& X, argument_type const& X1, result_type const& Y1, argument_type const& X2, result_type const& Y2, result_type& Result) const
+    result_type Interpolate(argument_type const& X, argument_type const& X1, result_type const& Y1, argument_type const& X2, result_type const& Y2) const
     {
-        const double epsilon = 1e-12;
-
-        double dx = X2 - X1;
-        result_type dy = Y2 - Y1;
+        // const double epsilon = 1e-12;
+        return Y1 + (X2 - X1 > 1.0e-12) * (Y2 - Y1)*(X - X1)/(X2 - X1);
```
Comment (pooyan-dadvand) [inline kratos/includes/table.h]: Any especial reason to take out the epsilon? You may make it constexp and leave it for better readability

File: kratos/includes/table.h (hunk 4 above)
Comment (pooyan-dadvand) [reply kratos/includes/table.h]: Also having the epsilon as constexp

**Review by pooyan-dadvand — COMMENTED** [review]
Just some minor comments!

File: kratos/includes/table.h [hunk 8]
```diff
@@ -617,21 +617,21 @@ class Table<double, double>
 
     result_type Interpolate(argument_type const& X, argument_type const& X1, result_type const& Y1, argument_type const& X2, result_type const& Y2) const
     {
-        // const double epsilon = 1e-12;
-        return Y1 + (X2 - X1 > 1.0e-12) * (Y2 - Y1)*(X - X1)/(X2 - X1);
+        constexpr double epsilon = 1.0e-12;
```
Comment (rubenzorrilla) [inline kratos/includes/table.h]: I think that what @pooyan-dadvand meant was to define the `epsilon` as a class `constexpr` so you ensure that a unique value is used.

File: kratos/python_scripts/run_cpp_tests.py [hunk 9]
```diff
@@ -1,3 +1,12 @@
-from KratosMultiphysics import *
-Tester.SetVerbosity(Tester.Verbosity.TESTS_OUTPUTS)
-Tester.RunAllTestCases()
+import KratosMultiphysics as Kratos
+import sys
+def run():
+    if len(sys.argv) < 2:
```
Comment (RiccardoRossi) [inline kratos/python_scripts/run_cpp_tests.py:4]: this change does not belong here...

File: kratos/includes/table.h [hunk 10]
```diff
@@ -603,24 +619,19 @@ class Table<double, double>
...
-
-        Result = Y1 + dy * scale;
-
-        return Result;
-
+        return Y1 + (X2 - X1 > epsilon) * ( (Y2 - Y1)*(X - X1)/(X2 - X1) );
+    }
+    result_type ExtrapolateBelowLowest(argument_type const& X, argument_type const& X1, result_type const& Y1, argument_type const& X2, result_type const& Y2) const
+    {
+        return Y1 - (X2 - X1 > epsilon) * ( (Y2 - Y1)*(X1 - X)/(X2 - X1 + std::numeric_limits<double>::min()) );
```
Comment (RiccardoRossi) [inline kratos/includes/table.h:628]: this is much less clear than the previous version

File: kratos/includes/table.h [hunk 11]
```diff
@@ -603,24 +619,19 @@ class Table<double, double>
...
     }
 
+    result_type ExtrapolateAboveHighest(argument_type const& X, argument_type const& X1, result_type const& Y1, argument_type const& X2, result_type const& Y2) const
+    {
+        return Y2 + (X2 - X1 > epsilon) * ( (Y2 - Y1)*(X - X2)/(X2 - X1 + std::numeric_limits<double>::min()) );
```
Comment (RiccardoRossi) [inline kratos/includes/table.h:633]: also.

and...i really cannot see why this should be faster than the previous version...

**Review by RiccardoRossi — CHANGES_REQUESTED** [review]
can u explain what is the key performance change, as i really cannot see it

File: kratos/includes/table.h (hunk 8 above)
Comment (pooyan-dadvand) [reply kratos/includes/table.h]: Wasn't my idea but I agree that it is better to have 1 epsilon for all

File: kratos/includes/table.h (hunk 10 above)
Comment (pooyan-dadvand) [reply kratos/includes/table.h:628]: @jrubiogonzalez do you have some performance comparison?

**Comment by jrubiogonzalez** [comment]
I did a performance comparison (for some kind of simulation accessing the tables is a major bottleneck) and tweaked the changes using VTune on a synthetic test that emulated our build (and therefore compiling in RelWithDebInfo).

image

For adding information to this dispute, I just created a test to show the differences, using 2 different tables (directly from one of our problems) accessing the whole range of the independent variables, inside and outside the table values.

I run the test, using 2 different computers and compilers compiling in Release (GCC on a desktop, VS on a laptop), my findings are:
For GCC proposed implementation is slightly faster (aprox 5-10% of the total time)
In VS the proposed implementation is a complete disaster and  consistently takes around twice the time.

So, IMHO I think it is not worth the change, I apologize for the noise.

I attach the results I got:

_GCC
TestPerformanceTable
0 C++ Time of tables: 9.96144 - 1.22515e+13
1 C++ Time of tables: 9.90343 - 1.22515e+13
2 C++ Time of tables: 9.89055 - 1.22515e+13
3 C++ Time of tables: 9.88366 - 1.22515e+13
4 C++ Time of tables: 9.89897 - 1.22515e+13
0 C++ Time of tables OldGetValue: 10.2688 - 1.22515e+13
1 C++ Time of tables OldGetValue: 10.4908 - 1.22515e+13
2 C++ Time of tables OldGetValue: 10.2066 - 1.22515e+13
3 C++ Time of tables OldGetValue: 10.6144 - 1.22515e+13
4 C++ Time of tables OldGetValue: 10.2491 - 1.22515e+13
 ------------------- Now with EnthalpyTables 
0 C++ Time of tables: 9.78924 - 3.05325e+18
1 C++ Time of tables: 9.7849 - 3.05325e+18
2 C++ Time of tables: 9.79238 - 3.05325e+18
3 C++ Time of tables: 9.76647 - 3.05325e+18
4 C++ Time of tables: 9.77588 - 3.05325e+18
0 C++ Time of tables OldGetValue: 10.5887 - 3.05325e+18
1 C++ Time of tables OldGetValue: 10.5941 - 3.05325e+18
2 C++ Time of tables OldGetValue: 10.5945 - 3.05325e+18
3 C++ Time of tables OldGetValue: 10.6187 - 3.05325e+18
4 C++ Time of tables OldGetValue: 10.5994 - 3.05325e+18_


_VS
TestPerformanceTable
0 C++ Time of tables: 18.6491 - 1.22515e+13
1 C++ Time of tables: 18.3574 - 1.22515e+13
2 C++ Time of tables: 18.8124 - 1.22515e+13
3 C++ Time of tables: 23.2141 - 1.22515e+13
4 C++ Time of tables: 22.3033 - 1.22515e+13
0 C++ Time of tables OldGetValue: 10.4174 - 1.22515e+13
1 C++ Time of tables OldGetValue: 11.4276 - 1.22515e+13
2 C++ Time of tables OldGetValue: 10.1061 - 1.22515e+13
3 C++ Time of tables OldGetValue: 10.4067 - 1.22515e+13
4 C++ Time of tables OldGetValue: 10.0959 - 1.22515e+13
 ------------------- Now with EnthalpyTables
0 C++ Time of tables: 29.2674 - 3.05325e+18
1 C++ Time of tables: 30.7856 - 3.05325e+18
2 C++ Time of tables: 31.2162 - 3.05325e+18
3 C++ Time of tables: 31.6102 - 3.05325e+18
4 C++ Time of tables: 29.7243 - 3.05325e+18
0 C++ Time of tables OldGetValue: 14.5903 - 3.05325e+18
1 C++ Time of tables OldGetValue: 14.2741 - 3.05325e+18
2 C++ Time of tables OldGetValue: 14.6618 - 3.05325e+18
3 C++ Time of tables OldGetValue: 15.0352 - 3.05325e+18
4 C++ Time of tables OldGetValue: 14.5527 - 3.05325e+18_

**Comment by rubenzorrilla** [comment]
@jrubiogonzalez I think it's very positive that we do and document this sort of tests (although the result is not as good as we could have expected in the beginning...)

Maybe the issue is the VS compiler optimizer, which AFAIK is not optimal at all...

Many thanks for the effort and the patience.

**Comment by loumalouomega** [comment]
I would keep this branch, so in the future we can retake this. One way to be sure that the optimizer does what we want is in fact doing ourselves. Many times optimization comes from decisions about branching that can be avoided if we avoid branching at all:

Page 43 and following: https://www.agner.org/optimize/optimizing_cpp.pdf

> The high speed of modern microprocessors is obtained by using a pipeline where
> instructions are fetched and decoded in several stages before they are executed. However,
> the pipeline structure has one big problem. Whenever the code has a branch (e.g. an if-else
> structure), the microprocessor does not know in advance which of the two branches to feed
> into the pipeline. If the wrong branch is fed into the pipeline then the error is not detected
> until 10 - 20 clock cycles later and the work it has done by fetching, decoding, and perhaps
> speculatively executing instructions during this time has been wasted. The consequence is
> that the microprocessor wastes several clock cycles whenever it feeds a branch into the
> pipeline and later discovers that it has chosen the wrong branch.

This can be avoided doing code which avoid branching, example:

~~~cpp
float a; bool b;
a = b ? 1.5f : 2.6f;
~~~

and 

~~~cpp
float a; bool b = 0;
const float lookup[2] = {2.6f, 1.5f};
a = lookup[b];
~~~

**Pull request closed without merging**